### ISSUE TITLE
Implement element size/position/visible event or a general read/write queue again

### ISSUE THREAD

**Issue by mhevery** [angular/angular#1693]
## Purpose

A component should be able to read its size and react to it. This is hard because we have to wait until DOM is fully rendered before causing layout and reading its position. (Additionally the browser does not have a resize event on elements)

## Mental Model

A component can register a `size` host event which the framework will call:

1. On initial render of the component to supply the initial size information.
2. When component size changes.

Component may change size due to:

1. Window resize
2. Changes to DOM rendering (ie text gets longer cause reflow.)

## Algorithm

1. dirty check all host elements which have `size` using `Element.getBoundingClientRect()` at the end of digest cycle when:
  * any DOM rendering occurs (may do pruning if we know that some things are absolutely position, later optimization)
  * a Window resize occurs. (orientation on mobile, which should cause a resize)
2. Invoke `size` handler only if the bounding rectangle has changed.
3. Schedule another `digest` before end of the turn. (This can be done be creating a `Promise` and resolving it within current `digest`.)

## Notes

* Other platforms may have this primitives and we don't have to dirty check
* Benefit of having it as event is that we can do proper batching of layouts.
* We can optimize (prune) dirty checking later once we know more about the application structure.
* `Ruler` Service can now be deleted.
* Need to split the `size` event and `position` concern into separate events. 
* Events: 
  * `size`: when width/height has changed. (Size changes less often then position)
  * `position`: when width/height or position changed.
  * `visible`: when the component is visible on the screen.

## Example

```javascript
@Component({
  selector: '...'
})
@View({
  template: `
  <ul style="overflow: scroll" 
        (scroll)="onScroll($event.target.scrollTop)"
        (size)="onResize($event)" >
    <li [style.height.px]="topPadding"></li>
    <li *for="var row in visibleRows"></li>
    <li [style.height.px]="bottomPadding"></li>
  </ul>
`
})
class VirtualScroll {
  visibleRows: Array;
  topPadding: number;
  buttomPadding: number;

  onScroll(offset: number) {
    // update the visibleRows
  }

  onResize(rect:Rectangle) {  
    // compute the number of visible rows.
  }
}
```

**Comment by tbosch** [comment]
Needed for migrating the GT calendar picker.

**Comment by kseamon** [comment]
On the surface, this seems reasonable. I know it's marked for later, but finding out of a dirtied bit of dom i out of flow might be more costly than just triggering the event naively.

Is it possible to trigger the event and subsequent dirty check synchronously (I assume using a promise makes it asynchronous)? If it's asynchronous, the user may perceive a flicker as the UI readjusts.

**Comment by tbosch** [comment]
@kseamon Yes, this should be possible to do without leaving the current vm turn (i.e. there should not be flicker). 

And maybe you are right that trying to reduce the number of elements that need to be checked might be slower than just checking all of them. Unless the application can give guarantees in some ways that don't rely on walking / checking the DOM at all.
E.g. a component could say that it is "encapsulated" regarding changing the position/size of elements that are not in its ShadowDom. I.e. if we detect that something inside of this component changed (via intercepting the renderer), we only need to dirty check the `size` listeners inside of this component and its children, but nowhere else.

This could actually be the default setting for all components...

**Comment by matanlurey** [comment]
I like the API. Will run by some engineers tomorrow for any additional input.

**Comment by tbosch** [comment]
I think we probably need a general debate of whether we want to have this event, or the DOM read/write queue back, based on concrete use cases. In GT, we should talk to isoos@...

Some use cases where GT uses the DOM read/write queue right now:

1. Third-party like charted need to know when angular is done with its work so not as to trigger additional reflows.
2. Some components need to read the size of the container so they know whether to load extra data to fill the components (ex: search showing 3 results in drop down but there is space for 5).
3. Add events that are enabled only after the component is displayed to avoid getting called during the rendering phase.

@mhevery What do you think?

**Comment by matanlurey** [comment]
My two cents:

- Keep a read/write queue option for advanced use cases. There will always be cases where you integrate with 3rd party libraries that will not be able to consume events, e.g., chart libraries, but still want to synchronize renders.

- Implement experimental events for size, position, visibility.

**Comment by tbosch** [comment]
@naomiblack I assigned this to M9, could you double check with Matan?

**Comment by matanlurey** [comment]
:+1: on M9

**Comment by tbosch** [comment]
FYI, GT has a prototype version of a DOM read/write queue in G3

**Comment by tbosch** [comment]
@mhevery @naomiblack GT is already implementing their `RulerService` and DOM read/write queue. As only they have the use cases for this so far, and it is not trivial to get describe in detail, let's not implement this in Angular for now and let GT drive it. Especially as we are going in the direction of [custom render services](https://github.com/angular/angular/issues/2409), with which users can just use a `RulerService` that returns an event stream instead of having a size event.

**Comment by tbosch** [comment]
As discussed offline, dropping prio, won't do it in this milestone.

**Comment by naomiblack** [comment]
This blocks internal b/20560018

**Comment by naomiblack** [comment]
This is internal issue b/21068639

**Comment by tbosch** [comment]
I would propose to close this as we decided to let GT drive this and once
their own version of the read/write queue is stabilized when copy it into
Angular.
